vid/src/avid_m/proofs.rs

```diff
@@ -0,0 +1,224 @@
+        param: &AvidMParam,
+        commit: &AvidMCommit,
+        shares: &[AvidMShare],
+    ) -> VidResult<MalEncodingProof> {
+        // First verify all the shares
+        for share in shares.iter() {
+            if AvidMScheme::verify_share(param, commit, share)?.is_err() {
+                return Err(VidError::Argument("Invalid share".to_string()));
+            }
+        }
+        // Recover the original payload in fields representation.
+        // Length of `payload` is always a multiple of `recovery_threshold`.
+        let witness = Self::recover_fields(param, shares)?;
+        let (mt, _) = Self::raw_encode(param, &witness)?;
+        if mt.commitment() == commit.commit {
+            return Err(VidError::Argument(
+                "Cannot generate the proof of incorrect encoding: encoding is good.".to_string(),
+            ));
+        }
+
+        let mut raw_shares = vec![];
+        let mut visited_indices = HashSet::new();
+        for share in shares {
+            for (index, mt_proof) in share
+                .content
+                .range
+                .clone()
+                .zip(share.content.mt_proofs.iter())
+            {
+                if index > param.total_weights {
+                    return Err(VidError::Argument("Invalid share".to_string()));
+                }
+                if visited_indices.contains(&index) {
+                    return Err(VidError::Argument("Overlapping shares".to_string()));
+                }
+                raw_shares.push(MalEncodingProofRawShare {
+                    index,
+                    mt_proof: mt_proof.clone(),
+                });
+                visited_indices.insert(index);
+                if raw_shares.len() >= param.recovery_threshold {
+                    break;
+                }
+            }
+        }
+        if raw_shares.len() < param.recovery_threshold {
+            return Err(VidError::Argument(
+                "Insufficient shares to generate the proof of incorrect encoding.".to_string(),
+            ));
+        }
+
+        Ok(MalEncodingProof {
+            witness,
+            raw_shares,
+        })
+    }
+}
+
+impl MalEncodingProof {
+    /// Verify a proof of incorrect encoding
+    pub fn verify(
+        &self,
+        param: &AvidMParam,
+        commit: &AvidMCommit,
+    ) -> VidResult<VerificationResult> {
+        // First check that all shares are valid.
+        if self.raw_shares.len() < param.recovery_threshold {
+            return Err(VidError::Argument(
+                "Insufficient shares to generate the proof of incorrect encoding.".to_string(),
+            ));
+        }
+        if self.raw_shares.len() > param.total_weights {
+            return Err(VidError::Argument("To many shares".to_string()));
+        }
+        let (mt, raw_shares) = AvidMScheme::raw_encode(param, &self.witness)?;
+        if mt.commitment() == commit.commit {
+            return Ok(Err(()));
```

Comment:
this is slightly confusing. (even tho i understand the intent) 

I'm not sure about the boundary between "invalid input" thus directly `Err()` and "valid input but failed verification, thus `Ok(Err())`. why isn't here also `Err(VidError::InvalidParam)`?

Reply:
Agree. Done in [4ccb13a](https://github.com/EspressoSystems/espresso-network/pull/2803/commits/4ccb13ae68f4f8ca7fbe573420f571a3309e0ffc)